### PR TITLE
fixes #28: allow duplicate property names except for __proto__

### LINKED ISSUE #28: ES6: remove duplicate object property (besides __proto__) restriction
### [B.3.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-__proto__-property-names-in-object-initializers) `__proto__` Property Names in Object Initializers

The following Early Error rule is added to those in 12.2.5.1:

    ObjectLiteral : { PropertyDefinitionList }

and

    ObjectLiteral : { PropertyDefinitionList , }

* It is a Syntax Error if PropertyNameList of PropertyDefinitionList contains any duplicate entries for `__proto__` and at least two of those entries were obtained from productions of the form `PropertyDefinition : PropertyName : AssignmentExpression`.

### PR DESCRIPTION
**Pull request by dwightmulcahy**

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -83,7 +83,7 @@ const BinaryPrecedence = {
   "-": Precedence.Additive,
   "*": Precedence.Multiplicative,
   "%": Precedence.Multiplicative,
-  "/": Precedence.Multiplicative,
+  "/": Precedence.Multiplicative
```
Comment (michaelficarra) [inline src/parser.js]: No.

File: src/parser.js [hunk 2]
```diff
@@ -1252,7 +1252,7 @@ export class Parser extends Tokenizer {
 
     if ({}.hasOwnProperty.call(propertyMap, key)) {
       if ((value & INIT_MASK) !== 0) {
-        if (this.strict && type === "DataProperty") {
+        if (this.strict && type === "DataProperty" || key == "$__proto__") {
```
Comment (michaelficarra) [inline src/parser.js]: `if (type === "DataProperty" && key == "$__proto__") {`

File: test/object-expressions/property-name.js [hunk 3]
```diff
@@ -21,5 +21,9 @@ describe("Parser", function () {
   describe("property name", function () {
     expect(ShiftParser.default("({0x0:0})").body.statements[0].expression.properties[0].name.value).to.be("0");
     expect(ShiftParser.default("({2e308:0})").body.statements[0].expression.properties[0].name.value).to.be("" + 1 / 0);
+
+    expect(function () { ShiftParser.default("({__proto__:42,__proto__:43})") }).to.throwError();
```
Comment (michaelficarra) [inline test/object-expressions/property-name.js]: Use our `assertParseFailure` assertion.

File: src/parser.js (hunk 1 above)
Comment (dwightmulcahy) [inline src/parser.js]: Bei said it *needs* to be there.

Comment (michaelficarra) [inline src/parser.js]: You *removed* it. I'm saying (like Bei) to add it back.

File: src/parser.js [hunk 4]
```diff
@@ -1252,7 +1252,7 @@ export class Parser extends Tokenizer {
 
     if ({}.hasOwnProperty.call(propertyMap, key)) {
       if ((value & INIT_MASK) !== 0) {
-        if (this.strict && type === "DataProperty") {
+        if (type === "DataProperty" && key == "$__proto__") {
```
Comment (michaelficarra) [inline src/parser.js]: Use `===`.

File: test/object-expressions/property-name.js [hunk 5]
```diff
@@ -16,10 +16,16 @@
...
+
 
 describe("Parser", function () {
   describe("property name", function () {
     expect(ShiftParser.default("({0x0:0})").body.statements[0].expression.properties[0].name.value).to.be("0");
     expect(ShiftParser.default("({2e308:0})").body.statements[0].expression.properties[0].name.value).to.be("" + 1 / 0);
+
+    assertParseFailure(function () { ShiftParser.default("({__proto__:42,__proto__:43})") });
+    assertParseFailure(function () { ShiftParser.default("use strict; ({__proto__:42,__proto__:43})") });
+    assertParseFailure(function () { ShiftParser.default("({__proto__:42},{__proto__:43})") });
```
Comment (michaelficarra) [inline test/object-expressions/property-name.js]: This is not the signature of `assertParseFailure`. It takes the invalid program and the expected error message, both as strings.

File: test/incompatibilities.js [hunk 6]
```diff
@@ -85,6 +85,13 @@ describe("Parser", function() {
       )
     );
 
+    // ES5: does not allow duplicate property names in strict mode
+    // ES6: allows duplicate property names except for __proto__
+    expect(stmt(parse("function hello() {\'use strict\'; ({ i: 42, i: 42 }) }",
```
Comment (michaelficarra) [inline test/incompatibilities.js]: This is not the signature of `expect`. These error messages shouldn't even be here.

**Comment by michaelficarra** [comment]
We need a failing test case for duplicate `__proto__` data properties. Also, a passing test case for getters/setters named `__proto__`.

File: test/object-expressions/property-name.js (hunk 5 above)
Comment (dwightmulcahy) [inline test/object-expressions/property-name.js]: removed these and added some success tests for __proto__, setter, and getter here.

File: test/incompatibilities.js (hunk 6 above)
Comment (dwightmulcahy) [inline test/incompatibilities.js]: removed these from here.

**Comment by dwightmulcahy** [comment]
Added success/failure tests for duplicate __proto__ and setter/getter.  
Renamed/reworded error msg associated with duplicate __proto__.

File: test/object-expressions/property-name.js [hunk 7]
```diff
@@ -21,5 +21,12 @@ describe("Parser", function () {
   describe("property name", function () {
     expect(ShiftParser.default("({0x0:0})").body.statements[0].expression.properties[0].name.value).to.be("0");
     expect(ShiftParser.default("({2e308:0})").body.statements[0].expression.properties[0].name.value).to.be("" + 1 / 0);
+    expect(ShiftParser.default("({__proto__:0})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
+    expect(ShiftParser.default("({get b() {}})").body.statements[0].expression.properties[0].name.value).to.be("b");
+    expect(ShiftParser.default("({set c(x) {}})").body.statements[0].expression.properties[0].name.value).to.be("c");
+    expect(ShiftParser.default("({get __proto__() {}})").body.statements[0].expression.properties[0].type).to.be("Getter");
+    expect(ShiftParser.default("({get __proto__() {}})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
+    expect(ShiftParser.default("({set __proto__(x) {}})").body.statements[0].expression.properties[0].type).to.be("Setter");
+    expect(ShiftParser.default("({set __proto__(x) {}})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
```
Comment (michaelficarra) [inline test/object-expressions/property-name.js]: We need a case with *both* a setter and a getter.

File: test/object-expressions/property-name.js [hunk 8]
```diff
@@ -21,5 +21,12 @@ describe("Parser", function () {
   describe("property name", function () {
     expect(ShiftParser.default("({0x0:0})").body.statements[0].expression.properties[0].name.value).to.be("0");
     expect(ShiftParser.default("({2e308:0})").body.statements[0].expression.properties[0].name.value).to.be("" + 1 / 0);
+    expect(ShiftParser.default("({__proto__:0})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
+    expect(ShiftParser.default("({get b() {}})").body.statements[0].expression.properties[0].name.value).to.be("b");
+    expect(ShiftParser.default("({set c(x) {}})").body.statements[0].expression.properties[0].name.value).to.be("c");
```
Comment (michaelficarra) [inline test/object-expressions/property-name.js]: These two are unrelated, but it doesn't hurt to have them. Move them up above the other `__proto__` tests.

Comment (dwightmulcahy) [inline test/object-expressions/property-name.js]: done.

File: test/object-expressions/property-name.js (hunk 7 above)
Comment (dwightmulcahy) [inline test/object-expressions/property-name.js]: done.

File: test/failure.js [hunk 9]
```diff
@@ -346,10 +342,8 @@ describe("Parser", function () {
     assertParseFailure("__proto__: __proto__: 42;", "Label \'__proto__\' has already been declared");
     assertParseFailure("\"use strict\"; function t(__proto__, __proto__) { }",
         "Strict mode function may not have duplicate parameter names");
-    assertParseFailure("\"use strict\"; x = { __proto__: 42, __proto__: 43 }",
-        "Duplicate data property in object literal not allowed in strict mode");
-    assertParseFailure("\"use strict\"; x = { get __proto__() { }, __proto__: 43 }",
-        "Object literal may not have data and accessor property with the same name");
+    assertParseFailure("({__proto__:1,__proto__:2})", "Duplicate __proto__ property in object literal not allowed");
+    assertParseFailure("{ \"use strict\"; ({__proto__:1,__proto__:2}) }", "Duplicate __proto__ property in object literal not allowed");
```
Comment (michaelficarra) [inline test/failure.js]: I forgot: add another test where one of them is quoted and the other isn't.

File: test/object-expressions/property-name.js [hunk 10]
```diff
@@ -21,5 +21,21 @@ describe("Parser", function () {
...
     expect(ShiftParser.default("({2e308:0})").body.statements[0].expression.properties[0].name.value).to.be("" + 1 / 0);
+
+    expect(ShiftParser.default("({get b() {}})").body.statements[0].expression.properties[0].name.value).to.be("b");
+    expect(ShiftParser.default("({set c(x) {}})").body.statements[0].expression.properties[0].name.value).to.be("c");
+
+    expect(ShiftParser.default("({__proto__:0})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
+
+    expect(ShiftParser.default("({get __proto__() {}})").body.statements[0].expression.properties[0].type).to.be("Getter");
+    expect(ShiftParser.default("({get __proto__() {}})").body.statements[0].expression.properties[0].name.value).to.be("__proto__");
+
```
Comment (michaelficarra) [inline test/object-expressions/property-name.js]: Instead of doubling up each of these tests, let's just compare the whole structure like we do in tests elsewhere. I can describe better in person.

Comment (dwightmulcahy) [inline test/object-expressions/property-name.js]: changed to comparing the structures...

File: test/failure.js (hunk 9 above)
Comment (dwightmulcahy) [inline test/failure.js]: added test for failure for {'__proto__':1,__proto__:2}
